capsules/src/bmp280.rs

```diff
@@ -0,0 +1,376 @@
+    state: Cell<State>,
+    /// Stores i2c commands
+    buffer: TakeCell<'static, [u8]>,
+    /// Needed to wait for readout completion, which can take milliseconds.
+    /// It's possible to implement this without an alarm with busy polling, but that's wasteful.
+    alarm: &'a A,
+}
+
+impl<'a, A: Alarm<'a>> Bmp280<'a, A> {
+    pub fn new(i2c: &'a dyn i2c::I2CDevice, buffer: &'static mut [u8], alarm: &'a A) -> Self {
+        Self {
+            i2c: I2cWrapper { i2c },
+            temperature_client: OptionalCell::empty(),
+            state: Cell::new(State::Uninitialized),
+            buffer: TakeCell::new(buffer),
+            alarm: alarm,
+        }
+    }
+
+    /// Resets the device and brings it into a known state.
+    pub fn begin_reset(&self) -> Result<(), ErrorCode> {
+        self.buffer.take().map_or_else(
+            || panic!("BMP280 No buffer available!"),
+            |buffer| match self.state.get() {
+                State::Uninitialized | State::Error => {
+                    self.i2c.read(buffer, Register::ID, 1);
+                    self.state.set(State::InitId);
+                    Ok(())
+                }
+                State::MismatchedDevice => Err(ErrorCode::NODEVICE),
+                _ => Err(ErrorCode::ALREADY),
+            },
+        )
+    }
+
+    pub fn read_temperature(&self) -> Result<(), ErrorCode> {
+        match self.state.get() {
+            // Actually, the sensor might be on, just in default state.
+            State::Uninitialized => Err(ErrorCode::OFF),
+            State::InitId
+            | State::InitResetting
+            | State::InitWaitingReady
+            | State::InitReadingCalibration => Err(ErrorCode::BUSY),
+            State::Idle(calibration) => {
+                self.buffer
+                    .take()
+                    .map_or(Err(ErrorCode::NODEVICE), |buffer| {
+                        // todo: use bitfield crate
+                        // forced mode, oversampling 1
+                        let val = 0b00100001;
+                        self.i2c.write(buffer, Register::CTRL_MEAS, [val]);
+                        self.state.set(State::Configuring(calibration));
+                        Ok(())
+                    })
+            }
+            State::Configuring(_)
+            | State::WaitingForAlarm(_)
+            | State::Waiting(_)
+            | State::Reading(_) => Err(ErrorCode::BUSY),
+            State::Error => Err(ErrorCode::FAIL),
+            State::MismatchedDevice => Err(ErrorCode::NODEVICE),
+        }
+    }
+
+    fn handle_alarm(&self) {
+        match self.state.get() {
+            State::WaitingForAlarm(calibration) => self.buffer.take().map_or((), |buffer| {
+                self.check_ready(buffer);
+                self.state.set(State::Waiting(calibration))
+            }),
+            other => {
+                debug!("BMP280 received unexpected alarm in state {:?}", other);
```

Comment:
Is there a compile-time switch to mark development builds? I would prefer to keep the debug messages at least when I'm testing. Comments aren't really a great way to do it.

Reply:
We do tend to avoid using conditional compiling in Tock, as this introduces issues related to testing. Using flags prevents all the code to be compiled in all cases and would require test builds for all the possible feature combinations.

Reply:
I opted for keeping the debug statements, improving the precedent of SHT3X. https://github.com/tock/tock/blob/master/capsules/src/sht3x.rs#L153

I haven't exercised a lot of the failure paths in the driver (and some outside), so I'm not confident enough to remove the flagging of bugs. If those early warnings are dropped silently, then finding the actual problems is going to be a real pain.

Reply:
The SHT3X driver is rather old and should not have those panics. 

My suggestion was to keep them there while you are debugging and comment them out when the driver is ready for production.

Reply:
I appreciate the suggestion, but I think there is a significant overlap between production and "not guaranteed bug free" (at least in things I build) and that I will never feel comfortable removing the debugs.

Reply:
It would be best to report this error to the caller instead of using a `debug!()` call

Reply:
I don't see any benefit of reporting this to the caller: this message is a compromise because I was asked to remove the panic, andit means "this driver (or OS) is buggy and does not function correctly". I struggle to see what the caller could do with this information.
This would also codify the driver being buggy as part of the driver API, which I haven't generally seen done, apart from the informal removing the resource and having the resource manager report a fatal error.
But Tock OS drivers don't seem to be making an allowance for themselves being buggy, so I think returning this to the caller would have been a departure.

Reply:
> But Tock OS drivers don't seem to be making an allowance for themselves being buggy, so I think returning this to the caller would have been a departure.

This is a subjective rather than authoritative response, but I do think that we do and should acknowledge that drivers can and will be buggy. Fundamentally, drivers being buggy is also a significant motivation of introducing the partly-unprivileged capsule concept in the first place. How to report such issues is a difficult question, but I would side with reporting it to userspace as well: when a capsule manages to misbehave without causing an overall denial of service, resilient userspace applications should have _the ability_ to recognize that and handle this gracefully, in which ever way they choose.  A good representation of a generic, unexpected capsule error is `ErrorCode::FAIL`.

I think it might be fine to also issue a console message on debug builds. We generally try to reduce error reporting in form of human-readable output -- at least for release builds -- because it tends to put significant strain on our flash space requirements (hence also favor `unwrap()` over `expect()`, etc.). Returning a generic internal failure to userspace also _enables_ userspace to report a human-readable error to the console, if it wishes to do so.